package/src/netmon.h

```diff
@@ -0,0 +1,75 @@
+// Copyright (C) CERN, 2018
+//
+// Network monitoring class
+//
+// This is a more efficient network monitor that holds
+// its istreams in a structure, so that they don't have
+// to be opened and closed on each monitoring cycle.
+
+#ifndef PRMON_NETMON_H
+#define PRMON_NETMON_H 1
+
+#include <fstream>
+#include <iostream>
+#include <memory>
+#include <string>
+#include <unordered_map>
+#include <vector>
+
+class netmon {
+ private:
+  // Which network interface paramters to measure
+  const std::vector<std::string> interface_params;
+
+  // Which network interfaces to monitor
+  std::vector<std::string> monitored_netdevs;
+
+  // Nested dictionary of network_if_streams[PARMETER][DEVICE][ISTREAM*]
+  std::unordered_map<
+      std::string,
+      std::unordered_map<std::string, std::unique_ptr<std::ifstream>>>
+      network_if_streams;
+
+  // Container for starting value stats, that are subtracted from measured
+  // values
+  std::unordered_map<std::string, unsigned long long> network_stats_start;
+
+  // Find all network interfaces on the system
+  std::vector<std::string> const get_all_network_devs();
+
+  // Open streams for each of the network interfaces and parameters
+  void open_interface_streams();
+
+  // Helper to map a device and parameter to a /sys filename
+  inline std::string const get_sys_filename(std::string device,
+                                            std::string param) {
+    std::string filename = "/sys/class/net/" + device + "/statistics/" + param;
+    return filename;
+  }
+
+  // Internal methods to read "raw" network stats
+  std::unordered_map<std::string, unsigned long long> read_raw_network_stats();
```

Comment:
Similar to the comment below. In this case we actually use this guy to store the initial values which are used as an offset. Shall we make it a `get` method that returns a `const` so that the result doesn't get changed mistakenly?